src/transformStateWithClones.js

```diff
@@ -7,7 +7,46 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    switch (actions[i].type) {
+      case 'addProperties':
+        if (i === 0) {
+          result.push(Object.assign({}, state, actions[i].extraData));
+          break;
+        } else {
+          result.push(Object.assign({}, result[i - 1], actions[i].extraData));
+        } break;
+
+      case 'removeProperties':
+        if (i === 0) {
+          const deleteProp = { ...state };
+
+          for (const key of actions[i].keysToRemove) {
+            delete deleteProp[key];
+          }
+
+          result.push(deleteProp);
+          break;
+        } else {
+          const deleteProp = { ...result[i - 1] };
+
+          for (const key of actions[i].keysToRemove) {
+            delete deleteProp[key];
+          }
+
+          result.push(deleteProp);
+          break;
+        }
+
```

Comment:
looks too complex, you can create the state object before the loop, and manipulate them in `for`
and at the end of the loop, only one time push that state object